File: backend/app/schemas/planner.py
from pydantic import BaseModel
from datetime import date, time
from typing import List, Optional

class SimulationRequest(BaseModel):
    semester_id: int
    dates: List[date]

class SubjectProjection(BaseModel):
    subject_id: int
    name: str
    code: Optional[str] = None
    current_percent: float
    projected_percent: float
    current_safe_bunks: int
    projected_safe_bunks: int
    is_safe: bool
    recovery_required: bool
    required_to_attend: int

class OverallProjection(BaseModel):
    current_percent: float
    projected_percent: float
    current_safe_bunks: int
    projected_safe_bunks: int

class MissedLectureInfo(BaseModel):
    subject_name: str
    date: date
    start_time: time
    end_time: time

class SimulationResponse(BaseModel):
    overall: OverallProjection
    subjects: List[SubjectProjection]
    missed_lectures: List[MissedLectureInfo]
    warnings: List[str]

class LeaveSuggestion(BaseModel):
    label: str
    start_date: date
    end_date: date
    dates: List[date]
    missed_classes_count: int
    projected_percent: float
    is_safe: bool

File: frontend/src/services/planner.ts
import api from "./api";

export interface SubjectProjection {
  subject_id: number;
  name: string;
  code?: string;
  current_percent: number;
  projected_percent: number;
  current_safe_bunks: number;
  projected_safe_bunks: number;
  is_safe: boolean;
  recovery_required: boolean;
  required_to_attend: number;
}

export interface OverallProjection {
  current_percent: number;
  projected_percent: number;
  current_safe_bunks: number;
  projected_safe_bunks: number;
}

export interface MissedLectureInfo {
  subject_name: string;
  date: string;
  start_time: string;
  end_time: string;
}

export interface SimulationResponse {
  overall: OverallProjection;
  subjects: SubjectProjection[];
  missed_lectures: MissedLectureInfo[];
  warnings: string[];
}

export interface LeaveSuggestion {
  label: string;
  start_date: string;
  end_date: string;
  dates: string[];
  missed_classes_count: number;
  projected_percent: number;
  is_safe: boolean;
}

export const plannerService = {
  async simulate(semesterId: number, dates: string[]): Promise<SimulationResponse> {
    const response = await api.post("/planner/simulate", {
      semester_id: semesterId,
      dates,
    });
    return response.data;
  },

  async getSuggestions(semesterId: number): Promise<LeaveSuggestion[]> {
    const response = await api.get("/planner/suggest", {
      params: { semester_id: semesterId },
    });
    return response.data;
  },
};
